File: src/easy/next-greater-element-i/1.js
/*
 * @lc app=leetcode id=496 lang=javascript
 *
 * [496] Next Greater Element I
 */

// @lc code=start
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function (nums1, nums2) {
  let recordMap = new Map();
  let stack = [];
  let result = [];
  for (let i = nums2.length - 1; i >= 0; i--) {
    while (stack.length > 0 && stack[stack.length - 1] < nums2[i]) {
      stack.pop();
    }
    recordMap.set(nums2[i], stack.length > 0 ? stack[stack.length - 1] : -1);
    stack.push(nums2[i]);
  }

  nums1.forEach((e) => {
    result.push(recordMap.get(e));
  });
  return result;
};
// @lc code=end

File: src/easy/valid-parentheses/1.go
package main

/*
 * @lc app=leetcode id=20 lang=golang
 *
 * [20] Valid Parentheses
 */

// @lc code=start

func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}

	m := map[string]string{
		"(": ")",
		"{": "}",
		"[": "]",
	}

	stack := make([]string, 0, len(s))

	for _, element := range s {
		if len(stack) > 0 {
			if stack[len(stack)-1] == m[string(element)] {
				stack = stack[:len(stack)-1]
			} else {
				stack = append(stack, (string(element)))
			}
		} else {
			stack = append(stack, string(element))
		}
	}
	return len(stack) == 0
}

// @lc code=end
